### PR TITLE
Add realtime flag for SCHED_RR

### PR DESCRIPTION
**Pull request by sad-goldfish**
#494

### CODE REVIEW

**Comment by emersion** [comment]
Do we need a new flag? I'd just change the priority unconditionally (just like we do for nice level).

We need to use `pthread_getschedparam` and `sched_get_priority_min` to properly set the new params. See https://github.com/swaywm/sway/pull/6994 for an example.

This PR has conflicts with the master branch.

**Comment by sad-goldfish** [comment]
@emersion The reason I added a new flag was because it makes the existing niceness infrastructure redundant. `SCHED_RR` processes do not respond to niceness. Rather than abruptly swapping these, a staged change seems more appropriate. And for `sched_get_priority_min`, on linux the minimum for `SCHED_RR` is 1. You can see this in `chrt -r -m` and `sched_get_priority_min(2)` though indeed changing this would get better compatibility with other unixes. Finally, I am not sure about `pthread_getschedparam`. Since we have `SCHED_RESET_ON_FORK` , if I understand correctly, we fork before exec'ing the game and hence that should reset to `SCHED_OTHER` right? (`sched(7)`).

The conflicts should be resolved with [c4996a2](https://github.com/Plagman/gamescope/pull/496/commits/c4996a216f3b48d74637f0d0cd337305e39574f8).

**Comment by sad-goldfish** [comment]
Ah, using `SCHED_RESET_ON_FORK` prevents the children from setting their niceness. Breaking the previous handling. I will fix this.

**Comment by sad-goldfish** [comment]
Looks good to me now.

**Comment by emersion** [comment]
We should `pthread_getschedparam` to fill the `struct sched_param` with current parameters, then overwrite the `sched_priority` field, then call `sched_setscheduler` with the updated struct. Otherwise unrelated parameters might be changed as well.

File: src/main.cpp [hunk 1]
```diff
@@ -373,6 +381,17 @@ int main(int argc, char **argv)
 			{
 				g_nNewNice = nNewNice;
 			}
+			if ( g_bRt )
+			{
+				int sched_prio = sched_get_priority_min(SCHED_RR);
+				struct sched_param sched = {.sched_priority = sched_prio};
+				if (pthread_getschedparam(pthread_self(), &g_nOldPolicy, &g_schedOldParam)) {
```
Comment (emersion) [inline src/main.cpp:388]: Nit: 0 can be used instead of `pthread_self` here

Comment (sad-goldfish) [reply src/main.cpp:388]: Are you sure? I get a segfault when I do this and `pthread_getschedparam(3)` does not say anything about this.

Comment (emersion) [reply src/main.cpp:390]: Hm, not sure where I found that, I can't find it anymore. Let's stick to `pthread_self` then.

Comment (emersion) [reply src/main.cpp:390]: Ah, I mixed it up with `sched_getparam`. `pthread_getschedparam` is different and doesn't accept 0.

**Review by emersion — APPROVED** [review]
Thanks!

**Comment by sad-goldfish** [comment]
I'm a bit new to this @emersion but I don't see anything wrong on my end. I think you can squash and merge the pull request on your end. If there is something wrong, could you point out where I should be looking or point me to some documentation for this?

**Comment by sad-goldfish** [comment]
Thank you for being patient @emersion. I think this is what you want.

**Pull request merged**